### PR TITLE
Add support for more pushable / non-pushable insert .. select queries with null-shard-key tables

### PR DESCRIPTION
**Pull request by onurctirtir**
* Add support for dist insert select by selecting from a reference table.
  
  This was the only pushable insert .. select case that
  #6773 didn't cover.

* For the cases where we insert into a Citus table but the INSERT .. SELECT
  query cannot be pushed down, allow pull-to-coordinator when possible.

  Remove the checks that we had at the very beginning of
  CreateInsertSelectPlanInternal so that we can try insert .. select via
  pull-to-coordinator for the cases where we cannot push-down the insert
  .. select query. What we support via pull-to-coordinator is still
  limited due to lacking of logical planner support for SELECT queries,
  but this commit at least allows using pull-to-coordinator for the cases
  where the select query can be planned via router planner, without
  limiting ourselves to restrictive top-level checks.

  Also introduce some additional restrictions into
  CreateDistributedInsertSelectPlan for the cases it was missing to check
  for null-shard-key tables. Indeed, it would make more sense to have
  those checks for distributed tables in general, via separate PRs against
  main branch. See https://github.com/citusdata/citus/pull/6817.

* Add support for inserting into a Postgres table.

### CODE REVIEW

File: src/backend/distributed/planner/insert_select_planner.c [hunk 1]
```diff
@@ -802,7 +744,18 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+		 * Note that we've already checked the non-existence of Postgres
+		 * tables in the subquery.
+		 */
+		if (subqueryRteListProperties->hasCitusLocalTable ||
+			subqueryRteListProperties->hasMaterializedView)
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "cannot select from a local relation when inserting "
+								 "into a distributed table that does not have a shard "
+								 "key", NULL, NULL);
```
Comment (onurctirtir) [inline src/backend/distributed/planner/insert_select_planner.c:757]: It would be nicer if we have this check for distributed tables in general, see https://github.com/citusdata/citus/pull/6817.

File: src/backend/distributed/planner/insert_select_planner.c [hunk 2]
```diff
@@ -737,6 +662,23 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+												targetRelationId);
+
+		if (!AllDistributedRelationsInListColocated(distributedRelationIdList))
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "distributed INSERT ... SELECT cannot reference a "
+								 "distributed table without a shard key together "
+								 "with non-colocated distributed tables",
+								 NULL, NULL);
+		}
```
Comment (onurctirtir) [inline src/backend/distributed/planner/insert_select_planner.c:679]: On 12.0, we should expand this to all distributed table types to avoid mistakenly pushing down insert .. select queries that reference non-colocated distributed tables in general. See example #1 in https://github.com/citusdata/citus/issues/6749.

File: src/test/regress/sql/insert_select_null_dist_key.sql [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+      SELECT a FROM distributed_table_c1_t1 WHERE b > 4
+    ) AS t3
+    JOIN (
+      SELECT a FROM distributed_table_c1_t2 WHERE b < 7
+    ) AS t4 ON t3.a = t4.a
+  ) AS t2
+);
+
+SET client_min_messages TO DEBUG2;
+
```
Comment (onurctirtir) [inline src/test/regress/sql/insert_select_null_dist_key.sql:338]: add tests with ON CONFLICT clause

File: src/test/regress/expected/insert_select_null_dist_key.out [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+-- TODO: this looks interesting
+INSERT INTO upsert_test_1 (unique_col, other_col) SELECT unique_col, other_col FROM upsert_test_1 ON CONFLICT (unique_col)
+	DO UPDATE SET other_col = (SELECT count(*) from upsert_test_1);
+DEBUG:  distributed statement: INSERT INTO insert_select_null_dist_key.upsert_test_1_1820106 AS citus_table_alias (unique_col, other_col) SELECT upsert_test_1.unique_col, upsert_test_1.other_col FROM insert_select_null_dist_key.upsert_test_1_1820106 upsert_test_1 ON CONFLICT(unique_col) DO UPDATE SET other_col = (SELECT count(*) AS count FROM insert_select_null_dist_key.upsert_test_1)
+ERROR:  cannot execute a distributed query from a query on a shard
+DETAIL:  Executing a distributed query in a function call that may be pushed to a remote node can lead to incorrect results.
+HINT:  Avoid nesting of distributed queries or use alter user current_user set citus.allow_nested_distributed_execution to on to allow it with possible incorrectness.
+CONTEXT:  while executing command on localhost:xxxxx
```
Comment (onurctirtir) [inline src/test/regress/expected/insert_select_null_dist_key.out:592]: This doesn't look correct, looking into that.

Comment (onurctirtir) [reply src/test/regress/expected/insert_select_null_dist_key.out:592]: Turns out that this is what we have on main too: https://github.com/citusdata/citus/issues/6826.

File: src/backend/distributed/planner/insert_select_planner.c [hunk 5]
```diff
@@ -819,25 +740,47 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+													targetRelationId);
+
+			if (!AllDistributedRelationsInListColocated(distributedRelationIdList))
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "distributed INSERT ... SELECT cannot reference a "
+									 "distributed table without a shard key together "
+									 "with non-colocated distributed tables",
+									 NULL, NULL);
+			}
```
Comment (onurctirtir) [inline src/backend/distributed/planner/insert_select_planner.c:783]: On 12.0, we should expand this to all distributed table types to avoid mistakenly pushing down insert .. select queries that reference non-colocated distributed tables in general. See example-1 in https://github.com/citusdata/citus/issues/6749.

File: src/backend/distributed/planner/insert_select_planner.c [hunk 6]
```diff
@@ -819,25 +740,49 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+			 * skip this check from the reference table case given that all reference tables
+			 * are already (and by default) co-located.
+			 */
+			if (!TablesColocated(insertRte->relid, selectPartitionColumnTableId))
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "INSERT target table and the source relation of the SELECT partition "
+									 "column value must be colocated in distributed INSERT ... SELECT",
+									 NULL, NULL);
+			}
```
Comment (onurctirtir) [inline src/backend/distributed/planner/insert_select_planner.c:785]: On 12.0, we should expand this to all distributed table types to avoid mistakenly pushing down insert .. select queries that reference non-colocated distributed tables in general. See example-1 in https://github.com/citusdata/citus/issues/6749.

File: src/backend/distributed/planner/insert_select_planner.c [hunk 7]
```diff
@@ -819,25 +740,49 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
...
+
+			if (!AllDistributedRelationsInListColocated(distributedRelationIdList))
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "distributed INSERT ... SELECT cannot reference a "
+									 "distributed table without a shard key together "
+									 "with non-colocated distributed tables",
+									 NULL, NULL);
+			}
 		}
```
Comment (onurctirtir) [inline src/backend/distributed/planner/insert_select_planner.c:762]: On 12.0, we should expand this to all distributed table types to avoid mistakenly pushing down insert .. select queries that reference non-colocated distributed tables in general. See example-1 in https://github.com/citusdata/citus/issues/6749.

File: src/test/regress/sql/insert_select_null_dist_key.sql [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+WHERE t2.sum_val > 2;
+
+-- MultiTaskRouterSelectQuerySupported() is unnecessarily restrictive
+-- about pushing down queries with DISTINCT ON clause even if the table
+-- doesn't have a shard key. See https://github.com/citusdata/citus/pull/6752.
+INSERT INTO nullkey_c1_t1 SELECT DISTINCT ON (a) a, b FROM nullkey_c1_t2;
+
+-- Similarly, we could push down the following query as well. see
+-- https://github.com/citusdata/citus/pull/6831.
+INSERT INTO nullkey_c1_t1 SELECT b, SUM(a) OVER (ORDER BY b) AS sum_val FROM nullkey_c1_t1;
```
Comment (onurctirtir) [inline src/test/regress/sql/insert_select_null_dist_key.sql:312]: Another thing that we might want to improve in general

File: src/test/regress/sql/insert_select_null_dist_key.sql [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+INSERT INTO nullkey_c1_t1 SELECT q.* FROM (SELECT reference_table.* FROM reference_table JOIN nullkey_c1_t1 USING (a)) q JOIN matview USING (a);
+
+-- use append / range distributed tables
+INSERT INTO nullkey_c1_t1 SELECT * FROM range_table;
+INSERT INTO nullkey_c1_t1 SELECT * FROM append_table;
+
+TRUNCATE nullkey_c1_t1, nullkey_c2_t1;
+INSERT INTO nullkey_c1_t1 SELECT i, i FROM generate_series(1, 8) i;
+INSERT INTO nullkey_c2_t1 SELECT i, i FROM generate_series(2, 7) i;
+
```
Comment (marcocitus) [inline src/test/regress/sql/insert_select_null_dist_key.sql:224]: could use some tests with RETURNING, ON CONFLICT, and maybe serial.

File: src/test/regress/expected/insert_select_null_dist_key.out [hunk 10]
```diff
@@ -620,10 +620,46 @@ INSERT INTO upsert_test_1 (unique_col, other_col) VALUES (1, 1) ON CONFLICT (uni
...
+INSERT INTO upsert_test_3 SELECT 7, other_col, 'harcoded_text_value' FROM upsert_test_1 RETURNING *;
+DEBUG:  distributed statement: INSERT INTO insert_select_null_dist_key.upsert_test_3_1820108 AS citus_table_alias (key_1, key_2, value) SELECT 7, upsert_test_1.other_col, 'harcoded_text_value'::text AS value FROM insert_select_null_dist_key.upsert_test_1_1820106 upsert_test_1 RETURNING citus_table_alias.key_1, citus_table_alias.key_2, citus_table_alias.value
+ key_1 | key_2 |        value
+---------------------------------------------------------------------
+     7 |     5 | harcoded_text_value
+(1 row)
+
+-- not yet supported
+INSERT INTO upsert_test_3 VALUES (1, 1, '2') ON CONFLICT (key_1, key_2) DO UPDATE SET key_2 = DEFAULT;
+ERROR:  queries that reference a distributed table without a shard key can only reference colocated distributed tables or reference tables
```
Comment (marcocitus) [inline src/test/regress/expected/insert_select_null_dist_key.out:661]: that's surprising, and seems unrelated to insert..select?

File: src/test/regress/sql/insert_select_null_dist_key.sql [hunk 11]
```diff
@@ -402,9 +402,21 @@ INSERT INTO upsert_test_1 (unique_col, other_col) VALUES (1, 1) ON CONFLICT (uni
...
 INSERT INTO upsert_test_1 (unique_col, other_col) VALUES (1, 1) ON CONFLICT (unique_col) WHERE unique_col = random()::int
 	DO UPDATE SET other_col = 5;
 
+CREATE TABLE upsert_test_3 (key_1 int, key_2 bigserial, value text DEFAULT 'default_value', PRIMARY KEY (key_1, key_2));
+SELECT create_distributed_table('upsert_test_3', null);
+
+INSERT INTO upsert_test_3 VALUES (1, DEFAULT, '1') RETURNING *;
+INSERT INTO upsert_test_3 VALUES (5, DEFAULT, DEFAULT) RETURNING *;
+INSERT INTO upsert_test_3 SELECT 7, other_col, 'harcoded_text_value' FROM upsert_test_1 RETURNING *;
+
```
Comment (marcocitus) [inline src/test/regress/sql/insert_select_null_dist_key.sql:416]: would be good to add an insert..select with on conflict as well

File: src/test/regress/expected/insert_select_null_dist_key.out (hunk 10 above)
Comment (onurctirtir) [reply src/test/regress/expected/insert_select_null_dist_key.out:661]: Right, this is actually not related to insert .. select.

And yes, it's because of not having support for logical planner yet.

File: src/test/regress/sql/insert_select_null_dist_key.sql (hunk 11 above)
Comment (onurctirtir) [reply src/test/regress/sql/insert_select_null_dist_key.sql:416]: Actually there are a few under `-- test upsert with INSERT .. SELECT queries`, do you think we need a few more?

File: src/test/regress/expected/insert_select_null_dist_key.out (hunk 10 above)
Comment (marcocitus) [reply src/test/regress/expected/insert_select_null_dist_key.out:662]: why would a single row upsert go through the logical planner?

File: src/test/regress/expected/insert_select_null_dist_key.out [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+ key_1 | key_2 |        value
+---------------------------------------------------------------------
+     7 |     5 | harcoded_text_value
+(1 row)
+
+SET client_min_messages TO DEBUG2;
+-- not yet supported
+INSERT INTO upsert_test_3 VALUES (1, 1, '2') ON CONFLICT (key_1, key_2) DO UPDATE SET key_2 = DEFAULT;
+ERROR:  queries that reference a distributed table without a shard key can only reference colocated distributed tables or reference tables
+DETAIL:  functions used in the DO UPDATE SET clause of INSERTs on distributed tables must be marked IMMUTABLE
```
Comment (onurctirtir) [inline src/test/regress/expected/insert_select_null_dist_key.out:663]: just remove that?

File: src/test/regress/expected/insert_select_null_dist_key.out (hunk 10 above)
Comment (onurctirtir) [reply src/test/regress/expected/insert_select_null_dist_key.out]: removed that test as we've discussed

File: src/test/regress/expected/insert_select_null_dist_key.out [hunk 13]
```diff
@@ -0,0 +1,720 @@
...
+DEBUG:  ALTER TABLE / ADD PRIMARY KEY will create implicit index "distributed_table_c1_t1_pkey" for table "distributed_table_c1_t1"
+DEBUG:  verifying table "distributed_table_c1_t1"
+INSERT INTO nullkey_c1_t1 AS t1 (a, b) SELECT t3.a, t3.b FROM nullkey_c1_t2 t2 JOIN reference_table t3 ON (t2.a = t3.a) ON CONFLICT (a)
+    DO UPDATE SET a = t1.a + 10;
+DEBUG:  distributed statement: INSERT INTO insert_select_null_dist_key.nullkey_c1_t1_1820000 AS t1 (a, b) SELECT t3.a, t3.b FROM (insert_select_null_dist_key.nullkey_c1_t2_1820001 t2 JOIN insert_select_null_dist_key.reference_table_1820003 t3 ON ((t2.a OPERATOR(pg_catalog.=) t3.a))) ON CONFLICT(a) DO UPDATE SET a = (t1.a OPERATOR(pg_catalog.+) 10)
+SET client_min_messages TO DEBUG1;
+INSERT INTO distributed_table_c1_t1 AS t1 (a, b) SELECT t3.a, t3.b FROM nullkey_c1_t2 t2 JOIN reference_table t3 ON (t2.a = t3.a) ON CONFLICT (a, b)
+    DO UPDATE SET b = t1.b + 10;
+DEBUG:  distributed INSERT ... SELECT cannot reference a distributed table without a shard key together with non-colocated distributed tables
+DEBUG:  Collecting INSERT ... SELECT results on coordinator
```
Comment (marcocitus) [inline src/test/regress/expected/insert_select_null_dist_key.out:698]: we should probably check the outcome of some of these (did we insert the right thing?)

**Review by marcocitus — APPROVED** [review]
Seems reasonable, we should fix the upsert separately.

File: src/test/regress/expected/insert_select_null_dist_key.out (hunk 13 above)
Comment (onurctirtir) [reply src/test/regress/expected/insert_select_null_dist_key.out:743]: Added some aggregate (avg) queries to check to rows that we've inserted into each table, at the end of particular test groups.

Verified that those avg values are correct by commenting out the calls made to create_distributed_table/create_reference_table, and by temporarily wrapping the commands that we throw an error for null-shard-key tables with BEGIN/ROLLBACK to make the comparison fair (because those commands do not fail for postgres tables ..).
